Insert channel funding outputs into Wallet

benthecarman opened the pull request:
When doing a splice, to properly calculate fees we need the channel's funding utxo in our wallet, otherwise our wallet won't know the channel's original size. This adds the channel funding txo to the wallet before a splice as we only really need this information for splices and not for all channels.

tnull commented on tests/integration_tests_rust.rs:
```diff
@@ -995,7 +995,7 @@ async fn splice_channel() {
 	node_b.splice_in(&user_channel_id_b, node_a.node_id(), 4_000_000).unwrap();
 
-	expect_splice_pending_event!(node_a, node_b.node_id());
+	let txo = expect_splice_pending_event!(node_a, node_b.node_id());
```
These test changes pass on `main`. Can you add coverage for the cases where the previous approach would lead to inaccurate fee estimations?

tnull reviewed (commented):
Makes sense to me. Do we think this is worth a backport for 0.7 or are we fine waiting for 0.8 with shipping this?

The different cases in the event handler could use a comment or two to explain when we'd land on which side of the `if`/`else` clauses there. Also, we really need to be careful with just replaying events in cases where we don't expect immediate resolution of the error on retry, as we'd otherwise just spin and block any other event processing. This could also use better test coverage to show why we need this change in the first place.

Besides that, also pinging @jkczyz as a second reviewer on this.

benthecarman commented on src/event.rs:
```diff
@@ -1458,6 +1458,33 @@ where
+					let chan_output = chans
+						.iter()
+						.find(|c| c.user_channel_id == user_channel_id)
+						.and_then(|c| c.get_funding_output());
```
> Doing this in ChannelReady wouldn't work for splicing preexisting channels.

I dont think this is right, in the docs it says we get a ChannelReady for splices, also my test seems to confirm

benthecarman replied:
ah i see for backward compat sake, added!

benthecarman commented on src/event.rs:
```diff
@@ -1458,6 +1458,33 @@ where
+							);
+						},
+						Some(output) => {
+							if let Err(e) = self.wallet.insert_txo(funding_txo, output) {
```
Made so we add it before a splice-in. I still think this one is good to keep so if you're doing any other wallet related action, it will have a more complete set of information and you don't need to initiate a splice for it to have the channel information

tnull replied:
> I still think this one is good to keep so if you're doing any other wallet related action, it will have a more complete set of information and you don't need to initiate a splice for it to have the channel information

Could you provide an example for what else we would use this information? If we don't need it I'd prefer to drop this change to declutter the logic here and avoid the extra persistence roundtrip.

jkczyz commented on src/lib.rs:
```diff
@@ -1267,29 +1267,25 @@ impl Node {
 			let shared_output = bitcoin::TxOut {
 				value: shared_input.previous_utxo.value + Amount::from_sat(splice_amount_sats),
-				script_pubkey: make_funding_redeemscript(&dummy_pubkey, &dummy_pubkey).to_p2wsh(),
+				script_pubkey: funding_output.script_pubkey,
```
This should still be the dummy one or note that it is not really the shared output's `script_pubkey`. I guess the amount is already fake since we don't know what the counterparty will contribute.

jkczyz commented on src/lib.rs:
```diff
@@ -1317,6 +1317,15 @@ impl Node {
+					log_error!(self.logger, "Failed to splice channel: {:?}", e);
+					Error::ChannelSplicingFailed
+				},
+			)?;
```
@tnull Don't know what the convention is here. Should we replace the `Error::PersistenceFailed` here with `Error::ChannelSplicingFailed`?

benthecarman commented:
Addressed review, ended up removing it on events and just doing it before a splice. Made sure test coverage hits on both backends and for splice in and out

tnull reviewed (approved):
LGTM, one nit. Otherwise should be ready pending @jkczyz's review.
